ubclaunchpad/foodies
# Backend ORM - TypeORM

Issue: https://github.com/ubclaunchpad/foodies/issues/35

_Summary_: 
Setup backend with ORM using TypeORM, this would be an alternative to raw SQL. I talked with @andrewzulaybar about exploring ORMs for our backend project, and we could go with this approach or SQL approach depending on how everyone feels. Just a heads up, I don't know too much about databases, I haven't taken 304. A lot of my terminology and logic may be incorrect, please correct me if somethings off.

I think in order to understand this PR, it may be best to go through the TypeORM documentation first, otherwise, a lot of things won't make sense. I don't know a lot about ORMs so I had to learn most of this from scratch, their documentation is pretty good.  https://typeorm.io/#/. I would also recommend checking out this branch, installing Postgres, creating a database called foodies, run yarn install, and then debug the project to see how everything works.

_What is an ORM?_ 
- https://en.wikipedia.org/wiki/Object%E2%80%93relational_mapping

_Why an ORM?_
This library makes it very easy for us to manage database migrations, leverage transactions, insert/update/delete from the database, and more. As you can see, we have a majority of our backend database schema mostly finished with little code. That way we can focus more so on our business logic and not have to worry about our project getting really large.

_Short summary:_
- Entities
  - https://github.com/ubclaunchpad/foodies/tree/backend-orm/backend/src/data
  - this folder contains all our entities, each entity is translated into a table in Postgres. As you can see by the annotations, there is a lot of customizable behaviour with respect to unidirectional/bidirectional associations, constraints, enums
- Repositories
  - each repository allows us to insert/update/delete entities.
https://github.com/ubclaunchpad/foodies/blob/408e111a8016ea047eefeca02a6849e5b906af6d/backend/src/App.ts#L33-L35
- Establishing a connection with Postgres
  - all we need is this line as well as `ormconfig.json`
https://github.com/ubclaunchpad/foodies/blob/408e111a8016ea047eefeca02a6849e5b906af6d/backend/src/App.ts#L12

_Special features_
- eager loading vs lazy loading
  - this is a very useful feature if we want to enhance performance
  - See https://typeorm.io/#/eager-and-lazy-relations
https://github.com/ubclaunchpad/foodies/blob/408e111a8016ea047eefeca02a6849e5b906af6d/backend/src/App.ts#L48-L56
- we can always use raw SQL if there are queries too complicated for typeORM, see https://typeorm.io/#/entity-manager-api
- different patterns in terms of how we access data
  - https://typeorm.io/#/active-record-data-mapper
- supported query builder features
  - https://typeorm.io/#/select-query-builder

_Database schema changes_
- I made some changes to the DB schema, we can definitely revert back to what is said in the doc, but these are some things I changed.
1. Instead of `User_uploads_promotion` join table
  - we now have a bidirectional one to many relationship where `promotion` is the owning side (so promotion has a column called `useID` and user has an array of `uploadedPromotions`.
  - since it's bidirectional, we can now see which user uploaded a certain promotion, and we can also see all the promotions a user uploaded
2. instead of `Promotion_has_discount` join table
  - we now have a bidirectional one to many relationship where `discount` is the owning side (so discount has a column called `promotionId`
  - now we can access all the discounts related to a promotion, as well as all promotions related to a discount (for search query purposes)
  - also not a many to many, see google doc comment https://docs.google.com/document/d/131JfzQXb472CyOL2yMSqbfuDyGI13UwQgnh0kowzPuk/edit
3.  Changed `discount_type` to an enum
4. `promotion_category` and `promotion_has_category`
  - I made all these enums, I think an enum makes more sense since we would be defining all possible types of categories that the user can select from. And we don't have to create another separate table for this
5.  Changed `cuisine_type` to enum
  - this is one of the harder enums since there can be lots of cuisine types, but I think it makes more sense to have an enum. This makes it easier to validate whether this is a cuisine type (so we prevent users from entering incorrect/ill-formatted cuisine types).
  - using enums is also good since the FE can maintain the same list of enums (or we can have an endpoint in BE returning all enums) and we don't have to waste a database transaction to query all the possible enums.


_Things I did not do for this PR_
- set max length of strings in the database
- make sure certain columns cannot be updated (`Promotion.dateAdded`, `Promotion.expirationDate`, etc.)
- upvote/downvote functionality
- restaurant table
- add validation logic
  - e.g. Discount.discountValue cannot have negative value, `expirationDate` has to be strictly greater than `dateAdded`
  - https://typeorm.io/#/validation
  - simple validation https://typeorm.io/#/decorator-reference/check
- did not finish configuring lazy/eager loading
- setup the migrations
- configure all indices

_Cons of TypeORM_
- when I was using `repository.insert([promotion1, promotion2, promotion3])` rather than individually saving each one, the cascades were not being saved correctly. This is a TypeORM bug I believe
  - **solution:** use `repository.save`
- when TypeORM has to create the tables from scratch, and then I call all the `repository.find()` methods, sometimes the repository returns the wrong entity. This is only the case when we have to create the tables from scratch (since `synchronize` is set to true in `ormconfig.json`, see below), but I believe once we arrive at a more concrete decision with our database schema, we won't be making as many drastic database schema changes
  - **solutions:** If we see this is happening too often, we can use the `QueryBuilder` rather than repositories to search for entities. We can also see if this behaviour keeps happening after our DB schema is finalized and we no longer need these two lines
https://github.com/ubclaunchpad/foodies/blob/408e111a8016ea047eefeca02a6849e5b906af6d/backend/ormconfig.json#L8-L9
  - there may be some more bugs related to `repository.find`, but these can be resolved by using the query builder
  - there may be some more bugs with cascade saves, I think I read somewhere this can be solved by using `create` and then `save`, or again, we can just use the raw SQL at the end.
With that said, I can't guarantee TypeORM will be bug-free, but I'm pretty sure these bugs have been discovered by many other people first, and we just have to research. With our use case as well, our schema is very tiny and not too complicated, so I don't think there will be that much trouble. But if TypeORM ends up being more of a trouble than raw SQL, and by then we're already using TypeORM, I can take the time to change our project to raw SQL.

_Things to be wary of:_
- https://typeorm.io/#/relations-faq/avoid-relation-property-initializers
- `ormconfig.json` can actually be a .env file if we want

_Cascade deletes_
- what we have configured right now (certainly not all the cases)
  - if you delete a user, this deletes all their uploaded promotions in the `Promotion` table and then deletes the promotions `discounts`. But if a user has a saved promotion that isn't uploaded by them, then that promotion won't be deleted
  - if you delete a promotion, you delete all the promotions discounts
  - deleting discount doesn't  delete promotion ( we should actually prohibit deletion of discounts unless the promotion is the one that triggered the deletion)
  - deleting promotion doesn't delete user (same reasoning above)
  - deleting promotion also removes any user's who reference the promotion in saved/uploaded.
